src/utils/SubscribableRequestQueue.ts

```diff
@@ -0,0 +1,175 @@
+    this.numSubscribers++;
+    this.subscribers.set(subscriberId, new Map());
+    return subscriberId;
+  }
+
+  /**
+   * Queues a new request, or adds a subscription if the request is already queued/running.
+   *
+   * If `subscriberId` is already subscribed to the request, this rejects the existing promise and returns a new one.
+   */
+  addRequestToQueue<T>(
+    key: string,
+    subscriberId: number,
+    requestAction: () => Promise<T>,
+    delayMs?: number
+  ): Promise<T> {
+    // Create single underlying request if it does not yet exist
+    if (!this.queue.hasRequest(key)) {
+      this.queue
+        .addRequest(key, requestAction, delayMs)
+        .then((value) => this.resolveAll(key, value))
+        .catch((reason) => this.rejectAll(key, reason));
+    }
+    if (!this.requests.has(key)) {
+      this.requests.set(key, []);
+    }
+
+    // Validate subscriber
+    if (subscriberId >= this.numSubscribers || subscriberId < 0) {
+      throw new Error(`SubscribableRequestQueue: subscriber id ${subscriberId} has not been registered`);
+    }
+    const subscriber = this.subscribers.get(subscriberId);
+    if (!subscriber) {
+      throw new Error(`SubscribableRequestQueue: subscriber id ${subscriberId} has been removed`);
+    }
+    const existingRequest = subscriber.get(key);
+    if (existingRequest) {
+      this.rejectSubscription(key, existingRequest, "SubscribableRequestQueue: request re-queued while running");
+    }
+
+    // Create promise and add to list of requests
+    return new Promise<T>((resolve, reject) => {
+      this.requests.get(key)?.push({ resolve, reject, subscriberId });
+      this.subscribers.get(subscriberId)?.set(key, reject);
+    });
+  }
+
+  /**
+   * Rejects a subscription and removes it from the list of subscriptions for a request, then cancels the underlying
+   * request if it is no longer subscribed and is not running already.
+   */
+  private rejectSubscription(key: string, reject: Rejecter, cancelReason?: unknown): void {
+    // Reject the outer "subscription" promise
+    reject(cancelReason);
+
+    // Get the list of subscriptions for this request
+    const subscriptions = this.requests.get(key);
+    if (!subscriptions) {
+      // This should never happen
+      return;
+    }
+    // Remove this request subscription by ref equality to `reject`
+    const idx = subscriptions.findIndex((sub) => sub.reject === reject);
+    if (idx >= 0) {
+      subscriptions.splice(idx, 1);
+    }
+
+    // Remove the underlying request if there are no more subscribers and the request is not already running
+    if (subscriptions.length < 1 && !this.queue.requestRunning(key)) {
+      this.queue.cancelRequest(key, cancelReason);
+      this.requests.delete(key);
+    }
+  }
+
+  /** Cancels a request subscription, and cancels the underlying request if it is no longer subscribed or running. */
+  cancelRequest(key: string, subscriberId: number, cancelReason?: unknown): void {
```

Comment:
is there any need for a cancelRequest that is independent of subscribers (e.g. `cancelForAllSubscribers(key, reason)`)?  Or does it always need to be initiated by an individual subscriber?

Reply:
I can't think of a use for one, but it would be simple enough to add since it just has to call the equivalent method on `RequestQueue`.

This does remind me of thoughts I've been having around how this class might relate to `RequestQueue`: would it be useful to allow `SubscribableRequestQueue` to be treated as a wrapper around `RequestQueue`? (i.e., allow passing in a `RequestQueue` to the constructor.) I could imagine, for example, using the bare `RequestQueue` for queueing prefetches on simple loaders, e.g. JSON time series frames, and only using the subscription abstraction for `OMEZarrLoader` (or any hypothetical future loader where data is chunked along more than one dimension).